### PR TITLE
Switch development to NixOS 21.11

### PR DESCRIPTION
**Pull request by dpausp**
@flyingcircusio/release-managers

### Significant breaking changes

* haproxy doesn't support `log localhost` anymore, use `log 127.0.0.1` instead
* PHP 5.6 is EOL for a long time and has been removed
* Postgresql 9.6 is EOL and has been removed
* OpenSSH 8.8 default settings break connections to old SSH daemons as ssh-rsa host keys and ssh-rsa public keys that were signed with SHA-1 are disabled now

### Possible problems

* iptables now uses nf_tables backend. We experienced problems when mixing legacy and new firewall rules that were gone after a reboot.

### Significant package updates

* bash now defaults to major version 5.
* python3 is now python39
* The varnish package was upgraded from 6.3.x to 7.x.
* k3s: 1.21 -> 1.22
* docker: 20.10.7 -> 20.10.9
* elasticsearch: 6.8.3 -> 6.8.21
* erlang: 23.3.4.5 -> 24.1.6
* docker-compose: 1.28.6 -> 1.29.2, 2.0.1
* git:  2.31.1 -> 2.33.1
* glibc: 2.32-54 ->  2.33-117
* kubernetes-helm: 3.5.4 -> 3.8.0
* imagemagick: 7.1.0-19 -> 7.1.0-24
* inetutils: 1.9.4 -> 2.2
* linux: 5.10.88 -> 5.10.99
* mailutils: 3.12 -> 3.13
* openssl: 1.1.1l -> 1.1.1m
* postfix: 3.6.0 -> 3.6.4
* rabbitmq: 3.8.9 -> 3.9.8
* sudo: 1.9.7 -> 1.9.7p2
* telegraf: 1.18.1 -> 1.20.3
* util-linux: 2.36.2 -> 2.37.3
* varnish: 6.3.2 -> 7.0.2
* systemd: 247 -> 249

## Other notable changes

* The setting services.openssh.logLevel "VERBOSE" "INFO". This brings NixOS in line with upstream and other Linux distributions, and reduces log spam on servers due to bruteforcing botnets.

## Release process

Impact:

Changelog:

(announce in changelog that there's a new platform version and main development has switched to it)

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - we want to follow NixOS releases to get upstream security fixes
  - look at release notes to find new security implications and possible breakage when upgrading systems
- [x] Security requirements tested? (EVIDENCE)
  - reviewed NixOS release notes for breaking changes, security-related and other significant changes 
  - all tests are green, various VMs already running on 21.11 without problems

### CODE REVIEW

File: nixos/platform/static.nix [hunk 1]
```diff
@@ -135,7 +135,6 @@ with lib;
 
     ids.uids = {
       # Our custom services
-      sensuserver = 31001;
       sensuapi = 31002;
```
Comment (ctheune) [inline nixos/platform/static.nix]: if sensu server is gone then the api and uchiwa need to go away, too

File: nixos/services/postgresql.nix [hunk 2]
```diff
@@ -56,7 +56,7 @@ in {
       majorVersion = mkOption {
           type = types.str;
           description = ''
-            The major version of PostgreSQL to use (9.6, 10, 11, 12, 13).
+            The major version of PostgreSQL to use (10, 11, 12, 13).
```
Comment (ctheune) [inline nixos/services/postgresql.nix]: 14 is missing here or make it more explicit that those are examples

File: pkgs/overlay.nix [hunk 3]
```diff
@@ -77,19 +62,7 @@ in {
...
-    version = "3.3.16";
-
-    src = fetchurl {
-      url = "https://packages.graylog2.org/releases/graylog/graylog-${version}.tgz";
-      sha256 = "17nxvj6haf5an6yj6zdjvcaxlliamcl16bca1z1jjcd7h9yjgxrz";
-    };
-  });
-
-  grub2_full = super.callPackage ./grub/2.0x.nix { };
+  #grub2_full = super.callPackage ./grub/2.0x.nix { };
```
Comment (ctheune) [inline pkgs/overlay.nix]: can be removed completely

File: release/default.nix [hunk 4]
```diff
@@ -95,6 +95,10 @@ let
   modifiedPkgNames = attrNames (import ../pkgs/overlay.nix pkgs pkgs);
 
   excludedPkgNames = [
+    # XXX: ceph doesn't build on 21.11
+    # Packages depending on ceph in fc are commented out in fc/default.nix
+    # because the exclusion mechanism here doesn't work recursively.
+    "ceph"
     # Build fails with patch errors.
     "gitlab"
```
Comment (ctheune) [inline release/default.nix]: i think gitlab can be removed here

File: release/default.nix [hunk 5]
```diff
@@ -95,6 +95,10 @@ let
   modifiedPkgNames = attrNames (import ../pkgs/overlay.nix pkgs pkgs);
 
   excludedPkgNames = [
+    # XXX: ceph doesn't build on 21.11
+    # Packages depending on ceph in fc are commented out in fc/default.nix
+    # because the exclusion mechanism here doesn't work recursively.
+    "ceph"
```
Comment (ctheune) [inline release/default.nix]: remove here so it becomes tested again when we enable the package

File: tests/default.nix [hunk 6]
```diff
@@ -80,12 +80,11 @@ in {
   postgresql11 = callTest ./postgresql.nix { rolename = "postgresql11"; };
   postgresql12 = callTest ./postgresql.nix { rolename = "postgresql12"; };
   postgresql13 = callTest ./postgresql.nix { rolename = "postgresql13"; };
-  postgresql96 = callTest ./postgresql.nix { rolename = "postgresql96"; };
+  postgresql14 = callTest ./postgresql.nix { rolename = "postgresql14"; };
   prometheus = callTest ./prometheus.nix {};
   rabbitmq = callTest ./rabbitmq.nix {};
   redis = callTest ./redis.nix {};
   rg-relay = callTest ./statshost/rg-relay.nix {};
-  sensu = callTest ./sensu.nix {};
```
Comment (ctheune) [inline tests/default.nix:88]: can we ensure that we have a test cover that the client still works?

**Review by ctheune — CHANGES_REQUESTED** [review]
see comments

File: tests/default.nix (hunk 6 above)
Comment (dpausp) [reply tests/default.nix:88]: added a new test that tests sensu client in isolation (just a local rabbitmq to connect to)

**Pull request merged**